mailbox/postgres/src/test/java/org/apache/james/mailbox/postgres/PostgresMailboxManagerTest.java

```diff
@@ -60,4 +82,159 @@ protected SubscriptionManager provideSubscriptionManager() {
+                .build(ClassLoaderUtils.getSystemResourceAsByteArray("eml/emailWithOnlyAttachment.eml")), session);
+
+            mailboxManager.deleteMailbox(inbox, session);
+
+            SoftAssertions.assertSoftly(softly -> {
+                PostgresMessageId messageId = (PostgresMessageId) appendResult.getId().getMessageId();
+                PostgresMailboxId mailboxId = (PostgresMailboxId) appendResult.getId().getMailboxId();
+
+                softly.assertThat(postgresMessageDAO.getBlobId(messageId).blockOptional())
+                    .isEmpty();
+
+                softly.assertThat(postgresMailboxMessageDAO.countTotalMessagesByMailboxId(mailboxId).block())
+                    .isEqualTo(0);
+            });
+        }
+
+        @Test
+        void deleteMessageInMailboxShouldEventuallyDeleteUnreferencedMessageMetadataWhenDeletingMessageFail() throws Exception {
+            PostgresMessageDAO spiedPostgresMessageDAO = spyPostgresMessageDAO();
+            doReturn(Mono.error(new RuntimeException("Fake exception")))
+                .doCallRealMethod()
+                .when(spiedPostgresMessageDAO).deleteByMessageId(Mockito.any());
+
+            MessageManager.AppendResult appendResult = inboxManager.appendMessage(MessageManager.AppendCommand.builder()
+                .build(ClassLoaderUtils.getSystemResourceAsByteArray("eml/emailWithOnlyAttachment.eml")), session);
+
+            inboxManager.delete(ImmutableList.of(appendResult.getId().getUid()), session);
+
+            SoftAssertions.assertSoftly(softly -> {
+                PostgresMessageId messageId = (PostgresMessageId) appendResult.getId().getMessageId();
+
+                softly.assertThat(postgresMessageDAO.getBlobId(messageId).blockOptional())
+                    .isEmpty();
+            });
+        }
+
+        private PostgresMailboxMessageDAO spyPostgresMailboxMessageDAO() throws Exception {
+            Field field = PostgresMailboxManagerProvider.deleteMessageListener.getClass().getDeclaredField("postgresMailboxMessageDAO");
+            field.setAccessible(true);
+            PostgresMailboxMessageDAO spiedPostgresMailboxMessageDAO = spy((PostgresMailboxMessageDAO) field.get(PostgresMailboxManagerProvider.deleteMessageListener));
+            field.set(PostgresMailboxManagerProvider.deleteMessageListener, spiedPostgresMailboxMessageDAO);
+            return spiedPostgresMailboxMessageDAO;
+        }
+
+        private PostgresMessageDAO spyPostgresMessageDAO() throws Exception {
+            Field field = PostgresMailboxManagerProvider.deleteMessageListener.getClass().getDeclaredField("postgresMessageDAO");
+            field.setAccessible(true);
+            PostgresMessageDAO spiedPostgresMessageDAO = spy((PostgresMessageDAO) field.get(PostgresMailboxManagerProvider.deleteMessageListener));
+            field.set(PostgresMailboxManagerProvider.deleteMessageListener, spiedPostgresMessageDAO);
+            return spiedPostgresMessageDAO;
+        }
+    }
```

Comment:
If going that way use at leadt `Mockito.spy()` directly on `PostgresMessageDAO` during construction?

I see coming this as a headeach during Java 21 migration...

Let's limit to the maximum direct use of reflection...

Reply:
```
        return postgresMailboxMessageDAO.deleteByMailboxId(mailboxId)
            .flatMap(this::handleMessageDeletion)
```
If in the first time, handleMessageDeletion fail, in the second time, because postgresMailboxMessageDAO has already deleted all corresponding records, there will be no messageId to feed handleMessageDeletion.

Reply:
Well, I would like us to work on another way to spy that messageDAO, in a way that we do not have custom injections.

Please have a look at this.

Reply:
sr I misunderstood you :)) fixed